### PR TITLE
[llvm-diff] Also diff alloca's allocated type and alignment

### PR DESCRIPTION
**Pull request by YanWQ-monad**
Currently, the function
https://github.com/llvm/llvm-project/blob/8c3304453c22ad1b5a914e64a7f6435f58f4099c/llvm/tools/llvm-diff/lib/DifferenceEngine.cpp#L364-L369
doesn't handle the non-operand data of `AllocaInst`. For example,
``` llvm
alloca i64, align 1
alloca i32, align 1
```
would be considered the same instruction.

---

Under some special circumstances, that would make llvm-diff crash:
``` llvm
define ptr @func() {
  %1 = alloca i64, align 1
  %2 = alloca i32, align 1
  %3 = alloca i8, align 1
  ret ptr %2
}
```

``` llvm
define ptr @func() {
  %1 = alloca i64, align 1
  %2 = alloca i32, align 1
  ret ptr %2
}
```

```
$ ./build/bin/llvm-diff lhs.ll rhs.ll
in function func:
  in block %0 / %0:
    in instruction   ret ptr %2 /   ret ptr %2:
      operands %2 and %2 differ
llvm-diff: path/llvm/tools/llvm-diff/lib/DifferenceEngine.cpp:268: void (anonymous namespace)::FunctionDifferenceEngine::unify(const Instruction *, const Instruction *): Assertion `!Result && "structural differences second time around?"' failed.
fish: Job 1, './build/bin/llvm-diff lhs.ll r…' terminated by signal SIGABRT (Abort)
```

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
This looks correct, but I think it only scratches the surface of the problem? The same problem exists for load types, GEP types, and basically everything else that is checked by `hasSameSpecialState()`. Can we use that instead?

**Comment by YanWQ-monad** [comment]
Adding test case

**Comment by YanWQ-monad** [comment]
When I was adding the test cases, I found another wired thing: same struct in different `.ll`s would be considered different types. This can also be seen from the failed tests.

I'm gonna investigate this issue tomorrow, because I have to go sleep now - I have lessons tomorrow.

**Comment by YanWQ-monad** [comment]
After investigation, I found that two `struct` declarations always have different `Type` pointers, despite they have the same element.

The related code might be
https://github.com/llvm/llvm-project/blob/ffe41819e58365dfbe85a22556c0d9d284e746b9/llvm/lib/IR/Type.cpp#L513-L529
We can infer that different calls to `StructType::create` will always create different `Type*`s, which cause `llvm-diff` (of the current PR) to consider same struct as different types.

So, should we modify the code in `Type.cpp` to make it generate same pointer for structure with same element, or handle this situation in `llvm-diff` in a special way so that it won't report the differences?

---

Reproduction (`llvm-diff` of the current PR):

``` llvm
%struct = type { i8 }

define void @gep(ptr %addr) {
  %1 = getelementptr %struct, ptr %addr, i32 0
  ret void
}
```

Command: `llvm-diff a.ll a.ll`, results in
```
in function gep:
  in block %0 / %0:
    >   %1 = getelementptr %struct.0, ptr %addr, i32 0
    <   %1 = getelementptr %struct, ptr %addr, i32 0
```

**Comment by YanWQ-monad** [comment]
TL;DR: added a function `isSameType(Type *TL, Type *TR)` to check whether `TL` on the left side and `TR` on the right side are structural equivalence. It recursively compares the inner elements of `StructType`, `VectorType`, and `ArrayType`, and caches the result in `IsSameTypes`.

In `diff(Instruction *L, Instruction *R, .....)` function, for type-related instructions (`load`, gep, `alloca`), check their type equivalence using `isSameType`; and for other instructions, it calls `hasSameSpecialState` to compare other non-operand data.

**Comment by YanWQ-monad** [comment]
ping?

**Comment by YanWQ-monad** [comment]
ping @nikic
